File: go/0300M-最长递增子序列.go
/*
 * @Author: SukiEva
 * @Date: 2021-12-06 08:56:19
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
package main

// * dp[i] 位置 i 的最长升序子序列
// - j从 0 到 i-1 各个位置的最长升序子序列 + 1 的最大值
func lengthOfLIS(nums []int) int {
	if len(nums) <= 1 {
		return len(nums)
	}
	dp := make([]int, len(nums))
	dp[0] = 1
	ans := 0
	for i := 1; i < len(nums); i++ {
		dp[i] = 1
		for j := 0; j < i; j++ {
			if nums[i] > nums[j] {
				dp[i] = max(dp[i], dp[j]+1)
			}
		}
		ans = max(ans, dp[i])
	}
	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: redo/0300M-最长递增子序列.cpp
/*
 * @Author: SukiEva
 * @Date: 2022-02-15 22:11:34
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        if (n <= 1) return n;
        vector<int> dp(n, 1);  // 严格单调增，类似单调栈，更新栈顶
        int ans = 0;
        for (int num : nums) {
            int l = 0, r = ans;
            while (l < r) {
                int m = (l + r) / 2;
                if (dp[m] < num)
                    l = m + 1;
                else
                    r = m;
            }
            dp[l] = num;
            if (ans == r) ans++;
        }
        return ans;
    }
};